File: python-backend/app/services/arbitrage.py
"""套利服务 - 核心业务逻辑封装"""
import logging
from typing import List, Tuple, Optional
from datetime import datetime

from app.core.config import Config
from app.core.models import (
    KalshiEvent, KalshiMarket,
    PolymarketEvent, PolymarketMarket,
    MatchedEvent, MatchedMarket,
    ArbitrageOpportunity,
    SystemStats
)
from app.core.matcher import EventMatcher
from app.core.calculator import ArbitrageCalculator
from app.clients.kalshi import KalshiClient
from app.clients.polymarket import PolymarketClient

logger = logging.getLogger(__name__)


class ArbitrageService:
    """套利服务 - 封装核心业务逻辑"""
    
    def __init__(self, config: Config):
        """初始化服务
        
        Args:
            config: 配置对象
        """
        self.config = config
        self.kalshi_client = KalshiClient(config.kalshi)
        self.polymarket_client = PolymarketClient(config.polymarket)
        self.matcher = EventMatcher()
        self.calculator = ArbitrageCalculator(
            min_profit_margin=config.settings.min_profit_margin,
            default_bet_amount=config.settings.default_bet_amount
        )
        
        # 数据存储
        self.kalshi_events: List[KalshiEvent] = []
        self.kalshi_markets: List[KalshiMarket] = []
        self.polymarket_events: List[PolymarketEvent] = []
        self.polymarket_markets: List[PolymarketMarket] = []
        self.matched_events: List[MatchedEvent] = []
        self.matched_markets: List[MatchedMarket] = []
        
        # 统计信息
        self.stats = SystemStats()
    
    async def initialize(self) -> bool:
        """初始化服务
        
        Returns:
            是否初始化成功
        """
        logger.info("=" * 60)
        logger.info("🚀 初始化套利服务")
        logger.info("=" * 60)
        
        # 登录 Kalshi
        logger.info("🔐 正在连接 Kalshi...")
        success = await self.kalshi_client.login()
        if not success:
            logger.error("❌ Kalshi 连接失败")
            return False
        
        logger.info("✅ Kalshi 连接成功")
        return True
    
    async def fetch_market_data(self) -> Tuple[int, int, int, int]:
        """获取市场数据
        
        Returns:
            (kalshi_events, kalshi_markets, poly_events, poly_markets) 数量元组
        """
        logger.info("=" * 60)
        logger.info("📥 开始获取市场数据")
        logger.info("=" * 60)
        
        # 获取 Kalshi 数据
        self.kalshi_events, self.kalshi_markets = await self.kalshi_client.get_nba_events_and_markets()
        
        # 获取 Polymarket 数据
        self.polymarket_events, self.polymarket_markets = await self.polymarket_client.get_nba_events_and_markets()
        
        # 更新统计
        self.stats.total_kalshi_events = len(self.kalshi_events)
        self.stats.total_kalshi_markets = len(self.kalshi_markets)
        self.stats.total_polymarket_events = len(self.polymarket_events)
        self.stats.total_polymarket_markets = len(self.polymarket_markets)
        
        logger.info(f"   Kalshi: {self.stats.total_kalshi_events} 事件, {self.stats.total_kalshi_markets} 市场")
        logger.info(f"   Polymarket: {self.stats.total_polymarket_events} 事件, {self.stats.total_polymarket_markets} 市场")
        
        return (
            self.stats.total_kalshi_events,
            self.stats.total_kalshi_markets,
            self.stats.total_polymarket_events,
            self.stats.total_polymarket_markets
        )
    
    async def match_markets(self) -> Tuple[int, int]:
        """匹配市场
        
        Returns:
            (matched_events, matched_markets) 数量元组
        """
        logger.info("=" * 60)
        logger.info("🔍 开始匹配市场")
        logger.info("=" * 60)
        
        # 执行两阶段匹配
        self.matched_events, self.matched_markets = self.matcher.match_events_and_markets(
            self.kalshi_events, self.kalshi_markets,
            self.polymarket_events, self.polymarket_markets
        )
        
        # 更新统计
        self.stats.matched_events = len(self.matched_events)
        self.stats.matched_markets = len(self.matched_markets)
        
        logger.info(f"   匹配: {self.stats.matched_events} 事件, {self.stats.matched_markets} 市场对")
        
        return (self.stats.matched_events, self.stats.matched_markets)
    
    def calculate_opportunities(self) -> List[ArbitrageOpportunity]:
        """计算套利机会
        
        Returns:
            套利机会列表
        """
        opportunities = []
        
        for mm in self.matched_markets:
            opp = self.calculator.calculate_single(
                event_name=mm.event_name,
                team_name=mm.team_name,
                kalshi_market=mm.kalshi_market,
                kalshi_yes_price=mm.kalshi_market.yes_price,
                kalshi_no_price=mm.kalshi_market.no_price,
                polymarket_market=mm.polymarket_market,
                polymarket_yes_price=mm.poly_yes_price,
                polymarket_no_price=mm.poly_no_price
            )
            
            if opp:
                opportunities.append(opp)
        
        # 按利润率排序
        opportunities.sort(key=lambda x: x.profit_margin, reverse=True)
        
        # 更新统计
        self.stats.arbitrage_opportunities = len(opportunities)
        self.stats.last_update = datetime.now()
        
        return opportunities
    
    def get_subscription_info(self) -> Tuple[List[str], List[str], dict]:
        """获取 WebSocket 订阅信息
        
        Returns:
            (kalshi_tickers, polymarket_token_ids, market_lookup) 元组
        """
        return self.matcher.get_subscription_info(self.matched_markets)
    
    async def scan_for_new_markets(self) -> Tuple[List[MatchedMarket], List[str], List[str], dict]:
        """扫描新市场，返回增量数据
        
        定时调用此方法检测新市场。只返回新增的匹配市场及其订阅信息，
        用于热订阅到现有 WebSocket 连接。
        
        Returns:
            (new_matched_markets, new_kalshi_tickers, new_poly_token_ids, new_market_lookup) 元组
            如果没有新市场，返回空列表和空字典
        """
        logger.info("=" * 60)
        logger.info("🔄 开始扫描新市场...")
        logger.info("=" * 60)
        
        # 1. 保存旧的已匹配市场 ID 集合
        old_matched_ids = set()
        for mm in self.matched_markets:
            # 使用 kalshi_market_id + team_name 作为唯一标识
            key = f"{mm.kalshi_market.market_id}_{mm.team_name}"
            old_matched_ids.add(key)
        
        old_count = len(self.matched_markets)
        old_kalshi_count = len(self.kalshi_markets)
        old_poly_count = len(self.polymarket_markets)
        logger.info(f"   扫描前状态:")
        logger.info(f"   - Kalshi: {old_kalshi_count} 个市场")
        logger.info(f"   - Polymarket: {old_poly_count} 个市场")
        logger.info(f"   - 已匹配: {old_count} 个市场对")
        
        # 2. 重新获取市场数据
        try:
            await self.fetch_market_data()
            new_kalshi_count = len(self.kalshi_markets)
            new_poly_count = len(self.polymarket_markets)
            
            kalshi_diff = new_kalshi_count - old_kalshi_count
            poly_diff = new_poly_count - old_poly_count
            
            logger.info(f"   扫描后状态:")
            logger.info(f"   - Kalshi: {new_kalshi_count} 个市场 ({'+' if kalshi_diff >= 0 else ''}{kalshi_diff})")
            logger.info(f"   - Polymarket: {new_poly_count} 个市场 ({'+' if poly_diff >= 0 else ''}{poly_diff})")
        except Exception as e:
            logger.error(f"❌ 获取市场数据失败: {e}")
            return [], [], [], {}
        
        # 3. 重新执行匹配
        try:
            await self.match_markets()
        except Exception as e:
            logger.error(f"❌ 匹配市场失败: {e}")
            return [], [], [], {}
        
        # 4. 找出新增的匹配市场
        new_matched_markets = []
        for mm in self.matched_markets:
            key = f"{mm.kalshi_market.market_id}_{mm.team_name}"
            if key not in old_matched_ids:
                new_matched_markets.append(mm)
        
        if not new_matched_markets:
            logger.info("   ✅ 没有发现新的匹配市场")
            logger.info("=" * 60)
            return [], [], [], {}
        
        logger.info(f"   🆕 发现 {len(new_matched_markets)} 个新匹配市场:")
        for i, mm in enumerate(new_matched_markets, 1):
            logger.info(f"      {i}. {mm.event_name} ({mm.team_name})")
        
        # 5. 生成新市场的订阅信息
        new_kalshi_tickers, new_poly_token_ids, new_market_lookup = \
            self.matcher.get_subscription_info(new_matched_markets)
        
        logger.info(f"   📡 新订阅需求:")
        logger.info(f"      - Kalshi: {len(new_kalshi_tickers)} 个市场")
        logger.info(f"      - Polymarket: {len(new_poly_token_ids)} 个 token")
        logger.info("=" * 60)
        
        return new_matched_markets, new_kalshi_tickers, new_poly_token_ids, new_market_lookup
    
    def get_stats(self) -> SystemStats:
        """获取统计信息
        
        Returns:
            系统统计信息
        """
        return self.stats
    
    async def close(self):
        """关闭服务"""
        if self.kalshi_client.session:
            await self.kalshi_client.session.close()
        if self.polymarket_client.session:
            await self.polymarket_client.session.close()

File: rust-backend/src/services/arbitrage.rs
//! Arbitrage Service
//!
//! Orchestrates market data fetching, matching, and arbitrage scanning.

use std::collections::HashSet;
use std::sync::Arc;
use std::time::{Duration, Instant};

use anyhow::Result;
use tokio::sync::mpsc;
use tracing::{error, info};

use crate::clients::{KalshiClient, PolymarketClient};
use crate::config::Config;
use crate::core::{EventMatcher, SubscriptionInfo};
use crate::models::{
    ArbitrageOpportunity, MatchedEvent, MatchedMarket, PolymarketPositionSide, PriceUpdate,
    SystemStats,
};
use crate::services::{ArbitrageStorage, Operation, PerformanceMetrics, WebSocketManager};

const MIN_REST_QUOTE_REFRESH_SECS: u64 = 5;

/// Arbitrage service
pub struct ArbitrageService {
    pub kalshi_client: KalshiClient,
    pub polymarket_client: PolymarketClient,
    pub matcher: EventMatcher,
    pub ws_manager: Arc<WebSocketManager>,
    pub storage: Arc<ArbitrageStorage>,
    pub matched_events: Vec<MatchedEvent>,
    pub matched_markets: Vec<MatchedMarket>,
    /// Performance metrics
    pub metrics: Arc<PerformanceMetrics>,
}

impl ArbitrageService {
    /// Create a new arbitrage service
    pub async fn new(config: &Config) -> Result<Self> {
        // Initialize storage
        let storage = Arc::new(ArbitrageStorage::new("arbitrage_history.db")?);

        // Initialize performance metrics
        let metrics = Arc::new(PerformanceMetrics::new());

        // Initialize clients
        let kalshi_client = KalshiClient::new(config.kalshi.clone())?;
        let mut polymarket_client = PolymarketClient::new(config.polymarket.clone())?;

        // Check the official Polymarket US order service for manual orders.
        if let Err(e) = polymarket_client.init_order_service().await {
            info!("Polymarket US order service initialization skipped: {}", e);
        }

        // Create matcher
        let matcher = EventMatcher::new(24);

        // Create WebSocket manager with metrics
        let mut ws_manager = WebSocketManager::new(
            config.settings.min_profit_margin,
            config.settings.default_bet_amount,
            config.settings.tracking_threshold,
            storage.clone(),
            metrics.clone(),
            Duration::from_secs(
                config
                    .settings
                    .refresh_interval
                    .max(MIN_REST_QUOTE_REFRESH_SECS)
                    .saturating_mul(2),
            ),
        );

        // Kalshi's live order book remains the only executable-depth source.
        ws_manager.set_kalshi_client(kalshi_client.clone());
        // Used only to fetch a one-time real depth snapshot when tracking
        // starts, for Advanced Search visibility; never for order submission.
        ws_manager.set_polymarket_client(polymarket_client.clone());

        // Load excluded markets from database
        ws_manager.load_excluded_markets();

        let ws_manager = Arc::new(ws_manager);

        Ok(Self {
            kalshi_client,
            polymarket_client,
            matcher,
            ws_manager,
            storage,
            matched_events: Vec::new(),
            matched_markets: Vec::new(),
            metrics,
        })
    }

    /// Initialize the service by fetching and matching markets
    pub async fn initialize(&mut self) -> Result<()> {
        info!("🔍 Fetching supported NBA and tennis match-winner data from both platforms...");

        // Fetch data from both platforms
        let (kalshi_events, kalshi_markets) = self
            .kalshi_client
            .get_supported_events_and_markets()
            .await?;

        let (polymarket_events, polymarket_markets) = self
            .polymarket_client
            .get_supported_events_and_markets()
            .await?;

        info!(
            "📊 Loaded supported sports: Kalshi {} events/{} markets, Polymarket {} events/{} markets",
            kalshi_events.len(),
            kalshi_markets.len(),
            polymarket_events.len(),
            polymarket_markets.len()
        );

        // Match events and markets (with timing)
        let match_start = Instant::now();
        let (matched_events, matched_markets) = self.matcher.match_events_and_markets(
            &kalshi_events,
            &kalshi_markets,
            &polymarket_events,
            &polymarket_markets,
        );
        self.metrics
            .record(Operation::MarketMatch, match_start.elapsed());

        self.matched_events = matched_events;
        self.matched_markets = matched_markets.clone();

        // Configure WebSocket manager
        self.ws_manager.set_matched_markets(matched_markets);
        let refreshed_polymarket_quotes = self
            .ws_manager
            .update_polymarket_rest_quotes(&polymarket_markets);
        let kalshi_tickers = self.ws_manager.get_kalshi_quote_tickers();
        let refreshed_kalshi_quotes =
            match self.kalshi_client.get_market_quotes(&kalshi_tickers).await {
                Ok(quotes) => self.ws_manager.update_kalshi_rest_quotes(&quotes),
                Err(error) => {
                    tracing::warn!(
                        "Initial authoritative Kalshi REST quote refresh failed: {}",
                        error
                    );
                    0
                }
            };

        info!(
            "✅ Initialization complete: {} matched markets, {} Kalshi and {} Polymarket REST quote updates",
            self.matched_markets.len(),
            refreshed_kalshi_quotes,
            refreshed_polymarket_quotes,
        );

        Ok(())
    }

    /// Start WebSocket connections for real-time updates
    pub async fn start_websocket_connections(
        &self,
        price_tx: mpsc::Sender<PriceUpdate>,
    ) -> Result<()> {
        let kalshi_tickers = self.ws_manager.get_kalshi_subscription_ids();

        info!(
            "📡 Starting WebSocket connections: {} Kalshi markets; Polymarket US uses REST quotes",
            kalshi_tickers.len()
        );

        let kalshi_client = self.kalshi_client.clone();

        let price_tx_kalshi = price_tx.clone();

        // Spawn Kalshi WebSocket
        let kalshi_tickers_clone = kalshi_tickers.clone();
        tokio::spawn(async move {
            loop {
                if let Err(e) = kalshi_client
                    .connect_websocket(kalshi_tickers_clone.clone(), price_tx_kalshi.clone())
                    .await
                {
                    error!(
                        "Kalshi WebSocket error: {}. Reconnecting in 5 seconds...",
                        e
                    );
                    tokio::time::sleep(tokio::time::Duration::from_secs(5)).await;
                }
            }
        });

        Ok(())
    }

    /// Run periodic market scanning
    pub async fn run_periodic_scan(&self, interval_secs: u64) {
        let ws_manager = self.ws_manager.clone();
        let interval_secs = interval_secs.max(MIN_REST_QUOTE_REFRESH_SECS);

        tokio::spawn(async move {
            let mut interval =
                tokio::time::interval(tokio::time::Duration::from_secs(interval_secs));

            loop {
                interval.tick().await;
                let opportunities = ws_manager.calculate_all();

                if !opportunities.is_empty() {
                    info!(
                        "📊 Periodic scan: found {} arbitrage opportunities, best: {:.2}%",
                        opportunities.len(),
                        opportunities
                            .first()
                            .map(|o| o.profit_margin)
                            .unwrap_or(0.0)
                    );
                }
            }
        });
    }

    /// Poll the Polymarket US gateway for current quotes on the same feed used
    /// by discovery. CLOB order books remain separate and are never inferred
    /// from the gateway's market-side IDs.
    pub async fn run_polymarket_quote_refresh(&self, requested_interval_secs: u64) {
        let interval_secs = requested_interval_secs.max(MIN_REST_QUOTE_REFRESH_SECS);
        if interval_secs != requested_interval_secs {
            tracing::warn!(
                "Polymarket quote refresh interval {}s is too low; clamping to {}s",
                requested_interval_secs,
                interval_secs
            );
        }

        let polymarket_client = self.polymarket_client.clone();
        let ws_manager = self.ws_manager.clone();

        tokio::spawn(async move {
            info!(
                "📡 Polymarket US REST quote refresh started, interval {} seconds",
                interval_secs
            );
            let mut interval =
                tokio::time::interval(tokio::time::Duration::from_secs(interval_secs));

            loop {
                interval.tick().await;

                match polymarket_client.get_supported_events_and_markets().await {
                    Ok((_, markets)) => {
                        let updated = ws_manager.update_polymarket_rest_quotes(&markets);
                        let matched_count = ws_manager.matched_markets.read().len();
                        if matched_count > 0 && updated == 0 {
                            tracing::warn!(
                                "Polymarket US REST quote refresh returned no quotes for {} matched markets",
                                matched_count
                            );
                        } else {
                            info!(
                                "✅ Polymarket US REST quote refresh updated {} matched markets",
                                updated
                            );
                        }
                    }
                    Err(e) => {
                        error!("❌ Polymarket US REST quote refresh failed: {}", e);
                    }
                }
            }
        });
    }

    /// Poll authoritative Kalshi REST asks for currently matched tickers.
    ///
    /// The WebSocket order-book cache stays independent and remains the only
    /// source allowed to satisfy execution-depth checks.
    pub async fn run_kalshi_quote_refresh(&self, requested_interval_secs: u64) {
        let interval_secs = requested_interval_secs.max(MIN_REST_QUOTE_REFRESH_SECS);
        if interval_secs != requested_interval_secs {
            tracing::warn!(
                "Kalshi quote refresh interval {}s is too low; clamping to {}s",
                requested_interval_secs,
                interval_secs
            );
        }

        let kalshi_client = self.kalshi_client.clone();
        let ws_manager = self.ws_manager.clone();

        tokio::spawn(async move {
            info!(
                "📡 Kalshi authoritative REST quote refresh started, interval {} seconds",
                interval_secs
            );
            let mut interval =
                tokio::time::interval(tokio::time::Duration::from_secs(interval_secs));

            loop {
                interval.tick().await;
                let tickers = ws_manager.get_kalshi_quote_tickers();
                if tickers.is_empty() {
                    continue;
                }

                match kalshi_client.get_market_quotes(&tickers).await {
                    Ok(quotes) => {
                        let updated = ws_manager.update_kalshi_rest_quotes(&quotes);
                        if updated != tickers.len() {
                            tracing::warn!(
                                "Kalshi REST quote refresh updated {}/{} matched tickers",
                                updated,
                                tickers.len()
                            );
                        } else {
                            info!(
                                "✅ Kalshi REST quote refresh updated {} matched tickers",
                                updated
                            );
                        }
                    }
                    Err(error) => {
                        error!("❌ Kalshi REST quote refresh failed: {}", error);
                    }
                }
            }
        });
    }

    /// Get current opportunities
    pub fn get_opportunities(&self) -> Vec<ArbitrageOpportunity> {
        self.ws_manager.get_opportunities()
    }

    /// Get system statistics
    pub fn get_stats(&self) -> SystemStats {
        let mut stats = self.ws_manager.get_stats();
        stats.matched_events = self.matched_events.len();
        stats.matched_markets = self.matched_markets.len();
        stats
    }

    /// Get matched markets
    pub fn get_matched_markets(&self) -> &[MatchedMarket] {
        &self.matched_markets
    }

    /// Place an order on Kalshi
    pub async fn place_kalshi_order(
        &self,
        ticker: &str,
        side: &str,
        outcome: &str,
        count: i32,
        price: i32,
    ) -> Result<serde_json::Value> {
        self.kalshi_client
            .place_order(ticker, side, outcome, count, price)
            .await
    }

    /// Place a Polymarket US market order.
    pub async fn place_polymarket_order(
        &self,
        market_slug: &str,
        position_side: PolymarketPositionSide,
        side: &str,
        contracts: i32,
        price: f64,
    ) -> Result<serde_json::Value> {
        self.polymarket_client
            .place_limit_order(market_slug, position_side, side, contracts, price)
            .await
    }

    /// Get arbitrage history
    pub fn get_arbitrage_history(
        &self,
        limit: usize,
    ) -> Result<Vec<crate::models::ArbitrageTrackingRecord>> {
        self.storage.get_history(limit)
    }

    /// Scan for new markets and return incremental subscription info
    ///
    /// This method fetches fresh market data from both platforms, runs matching,
    /// and returns only the NEW matched markets that weren't in the previous set.
    ///
    /// Returns: (new_matched_markets, new_subscription_info)
    pub async fn scan_for_new_markets(&mut self) -> Result<(Vec<MatchedMarket>, SubscriptionInfo)> {
        info!("============================================================");
        info!("🔄 Starting scan for new markets...");
        info!("============================================================");

        // 1. Save old matched market IDs
        let old_matched_ids: HashSet<String> = self
            .matched_markets
            .iter()
            .map(|mm| format!("{}_{}", mm.kalshi_market.market_id, mm.team_name))
            .collect();

        let old_count = self.matched_markets.len();
        info!("   Scan state before scan: {} matched markets", old_count);

        // 2. Fetch fresh market data
        let (kalshi_events, kalshi_markets) =
            match self.kalshi_client.get_supported_events_and_markets().await {
                Ok(data) => data,
                Err(e) => {
                    error!("❌ Failed to fetch Kalshi market data: {}", e);
                    return Ok((Vec::new(), SubscriptionInfo::empty()));
                }
            };

        let (polymarket_events, polymarket_markets) = match self
            .polymarket_client
            .get_supported_events_and_markets()
            .await
        {
            Ok(data) => data,
            Err(e) => {
                error!("❌ Failed to fetch Polymarket market data: {}", e);
                return Ok((Vec::new(), SubscriptionInfo::empty()));
            }
        };

        info!(
            "   Scan state after scan: Kalshi {} events/{} markets, Polymarket {} events/{} markets",
            kalshi_events.len(),
            kalshi_markets.len(),
            polymarket_events.len(),
            polymarket_markets.len()
        );

        // 3. Re-run matching
        let match_start = Instant::now();
        let (matched_events, matched_markets) = self.matcher.match_events_and_markets(
            &kalshi_events,
            &kalshi_markets,
            &polymarket_events,
            &polymarket_markets,
        );
        self.metrics
            .record(Operation::MarketMatch, match_start.elapsed());

        // 4. Find new matched markets
        let mut new_matched_markets = Vec::new();
        for mm in &matched_markets {
            let key = format!("{}_{}", mm.kalshi_market.market_id, mm.team_name);
            if !old_matched_ids.contains(&key) {
                new_matched_markets.push(mm.clone());
            }
        }

        // 5. Update internal state
        self.matched_events = matched_events;
        self.matched_markets = matched_markets;

        if new_matched_markets.is_empty() {
            info!("   ✅ No new matched markets discovered");
            info!("============================================================");
            return Ok((Vec::new(), SubscriptionInfo::empty()));
        }

        info!("   🆕 发现 {} 个新匹配市场:", new_matched_markets.len());
        for (i, mm) in new_matched_markets.iter().enumerate().take(5) {
            info!("      {}. {} ({})", i + 1, mm.event_name, mm.team_name);
        }
        if new_matched_markets.len() > 5 {
            info!(
                "      ... {} more new markets",
                new_matched_markets.len() - 5
            );
        }

        // 6. Generate subscription info for new markets only
        let new_sub_info = self.matcher.get_subscription_info(&new_matched_markets);

        info!(
            "   📡 New subscription requirements: Kalshi {} markets; Polymarket US uses REST quotes",
            new_sub_info.kalshi_tickers.len()
        );
        info!("============================================================");

        Ok((new_matched_markets, new_sub_info))
    }

    /// Search history records with filters
    pub fn search_history(
        &self,
        min_profit: Option<f64>,
        max_profit: Option<f64>,
        min_duration: Option<f64>,
        max_duration: Option<f64>,
        event_name: Option<String>,
        team_name: Option<String>,
        sort_by: Option<String>,
        sort_order: Option<String>,
        limit: Option<usize>,
        offset: Option<usize>,
        include_history: Option<bool>,
    ) -> Result<serde_json::Value> {
        self.storage.search_records(
            min_profit,
            max_profit,
            min_duration,
            max_duration,
            event_name,
            team_name,
            sort_by,
            sort_order,
            limit,
            offset,
            include_history,
        )
    }

    /// Get history statistics
    pub fn get_history_statistics(&self) -> Result<serde_json::Value> {
        self.storage.get_statistics()
    }
}
